pkg/pool-weighted/contracts/LBPool.sol

```diff
@@ -0,0 +1,294 @@
+        }
+        if (endWeights[0] + endWeights[1] != FixedPoint.ONE) {
+            revert NormalizedWeightInvariant();
+        }
+
+        // Ensure startTime >= now.
+        startTime = GradualValueChange.resolveStartTime(startTime, endTime);
+
+        // The SafeCast ensures `endTime` can't overflow.
+        _startGradualWeightChange(startTime.toUint32(), endTime.toUint32(), _getNormalizedWeights(), endWeights);
+    }
+
+    /*******************************************************************************
+                                    Hook Functions
+    *******************************************************************************/
+
+    /**
+     * @notice Hook to be executed when pool is registered.
+     * @dev Returns true if registration was successful, and false to revert the registration of the pool.
+     * @param pool Address of the pool
+     * @return success True if the hook allowed the registration, false otherwise
+     */
+    function onRegister(
+        address,
+        address pool,
+        TokenConfig[] memory,
+        LiquidityManagement calldata
+    ) public view override onlyVault returns (bool) {
+        // Since in this case the pool is the hook, we don't need to check anything else.
+        // We *could* check that it's two tokens, but better to let that be caught later, as it will fail with a more
+        // descriptive error.
+        return pool == address(this);
+    }
+
+    // Return HookFlags struct that indicates which hooks this contract supports
+    function getHookFlags() public pure override returns (HookFlags memory hookFlags) {
+        // Check whether swaps are enabled in `onBeforeSwap`.
+        hookFlags.shouldCallBeforeSwap = true;
+        // Ensure the caller is the owner, as only the owner can add liquidity.
+        hookFlags.shouldCallBeforeAddLiquidity = true;
+    }
+
+    /**
+     * @notice Check that the caller who initiated the add liquidity operation is the owner.
+     * @dev We first ensure the caller is the standard router, so that we know we can trust the value it returns
+     * from `getSender`.
+     *
+     * @param router The address (usually a router contract) that initiated the add liquidity operation
+     */
+    function onBeforeAddLiquidity(
+        address router,
+        address,
+        AddLiquidityKind,
+        uint256[] memory,
+        uint256,
+        uint256[] memory,
+        bytes memory
+    ) public view override onlyVault returns (bool) {
+        if (router == _TRUSTED_ROUTER) {
+            return IRouterCommon(router).getSender() == owner();
+        }
+
+        revert RouterNotTrusted();
+    }
+
+    /**
+     * @notice Called before a swap to let the pool block swaps if not enabled.
+     * @return success True if the pool has swaps enabled.
+     */
+    function onBeforeSwap(PoolSwapParams calldata, address) public view override onlyVault returns (bool) {
+        return _getPoolSwapEnabled();
```

Comment:
More than just the overhead of calling the hook itself, this is mainly because it needs to reload balances and rates after calling a before hook, since they might have changed: even though all we're doing here is checking a flag. We'd briefly considered having before hooks return whether they need to be reloaded, but that's asking for trouble (a malicious one could say they didn't need a reload when they did).

Reply:
I don't have a strong preference either way on this one. The hook seemed most elegant, but overriding `onSwap` seems easy enough. Realistically I'm going to need to override `onSwap` for the seedless version anyway, so that's fine with me

Reply:
It would add significant gas to every swap - thousands at least - with all the reloading. It's the side effects that make it expensive: and aren't apparent until you look at the Vault code. If you're going to do it anyway, I think it makes sense to introduce that now.